spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/KubernetesDiscoveryClient.java

```diff
@@ -107,34 +107,29 @@ public List<ServiceInstance> getInstances(String serviceId) {
 				.map(KubernetesDiscoveryClientUtils::subsetsFromEndpoints).toList();
 
 		List<ServiceInstance> instances = new ArrayList<>();
-		if (!subsetsNS.isEmpty()) {
-			for (EndpointSubsetNS es : subsetsNS) {
-				instances.addAll(getNamespaceServiceInstances(es, serviceId));
-			}
+		for (EndpointSubsetNS es : subsetsNS) {
+			instances.addAll(getNamespaceServiceInstances(es, serviceId));
 		}
 
 		return instances;
 	}
 
 	public List<Endpoints> getEndPointsList(String serviceId) {
 		if (properties.allNamespaces()) {
-			return client.endpoints().inAnyNamespace().withField("metadata.name", serviceId)
```

Comment:
first, I inlined `findEndPointsFilteredByNamespaces` into `getEndPointsList`, so the code became a bit simpler to read:

```
	public List<Endpoints> getEndPointsList(String serviceId) {
		if (properties.allNamespaces()) {
			LOG.debug(() -> "searching for endpoints in all namespaces");
			return client.endpoints().inAnyNamespace().withField("metadata.name", serviceId)
					.withLabels(properties.serviceLabels()).list().getItems();
		}
		else if (properties.namespaces().isEmpty()) {
			LOG.debug(() -> "searching for endpoints in namespace : " + client.getNamespace());
			return client.endpoints().withField("metadata.name", serviceId).withLabels(properties.serviceLabels())
					.list().getItems();
		}
		else {
			LOG.debug(() -> "searching for endpoints in namespaces : " + properties.namespaces());
			List<Endpoints> endpoints = new ArrayList<>();
			for (String namespace : properties.namespaces()) {
				endpoints.addAll(client.endpoints().inNamespace(namespace).withField("metadata.name", serviceId)
					.withLabels(properties.serviceLabels()).list().getItems());
			}
			return endpoints;
		}
	}
```


Then I noticed that we re-use this `withField("metadata.name", serviceId)
					.withLabels(properties.serviceLabels()).list().getItems()` in all 3 branches, so I had to find a way to extract it. I re-used an addition in the fabric8 client called `FilterNested`, basically you can do:

```
FilterNested<FilterWatchListDeletable<Endpoints, EndpointsList, Resource<Endpoints>>> filterNested = client.endpoints().inAnyNamespace().withNewFilter();
```

and then chain the filters you want. So I created a static method:

```
	static List<Endpoints> endpoints(
			FilterNested<FilterWatchListDeletable<Endpoints, EndpointsList, Resource<Endpoints>>> filterNested,
			KubernetesDiscoveryProperties properties, String serviceId) {
		return filterNested.withField("metadata.name", serviceId).withLabels(properties.serviceLabels()).endFilter()
				.list().getItems();
	}
```

in `KubernetesDiscoveryClientUtils` and refactored the code to:

```
	public List<Endpoints> getEndPointsList(String serviceId) {
		if (properties.allNamespaces()) {
			LOG.debug(() -> "searching for endpoints in all namespaces");
			return endpoints(client.endpoints().inAnyNamespace().withNewFilter(), properties, serviceId);
		}
		else if (properties.namespaces().isEmpty()) {
			LOG.debug(() -> "searching for endpoints in namespace : " + client.getNamespace());
			return endpoints(client.endpoints().withNewFilter(), properties, serviceId);
		}
		else {
			LOG.debug(() -> "searching for endpoints in namespaces : " + properties.namespaces());
			List<Endpoints> endpoints = new ArrayList<>();
			for (String namespace : properties.namespaces()) {
				endpoints.addAll(
						endpoints(client.endpoints().inNamespace(namespace).withNewFilter(), properties, serviceId));
			}
			return endpoints;
		}
	}
```

I've also adjusted some failing tests and added some more.